frontend/src/metabase/admin/datamodel/containers/MetricApp.jsx

```diff
@@ -6,73 +6,86 @@ import MetabaseAnalytics from "metabase/lib/analytics";
+  wrapped: true,
+})
+@Tables.load({ id: (state, props) => props.metric.table_id, wrapped: true })
+class UpdateMetricForm extends Component {
+  constructor() {
+    super();
+    this.onSubmit = this.onSubmit.bind(this);
+  }
 
-    let tableId;
-    if (params.id) {
-      const metricId = parseInt(params.id);
-      const { payload: metric } = await this.props.getMetric({ metricId });
-      tableId = metric.table_id;
-    } else if (location.query.table) {
-      tableId = parseInt(location.query.table);
-    }
+  async onSubmit(metric) {
+    await this.props.metric.update(metric);
+    MetabaseAnalytics.trackEvent("Data Model", "Metric Updated");
+    const { id: tableId, db_id: databaseId } = this.props.table;
+    this.props.onChangeLocation(
+      `/admin/datamodel/database/${databaseId}/table/${tableId}`,
+    );
+  }
 
-    if (tableId != null) {
-      // TODO Atte Keinänen 6/8/17: Use only global metadata (`fetchTableMetadata`)
-      this.props.loadTableMetadata(tableId);
-      this.props.fetchTableMetadata(tableId);
+  render() {
+    if (this.props.table) {
+      this.props.table.fetchTableMetadata();
     }
+    return <MetricForm {...this.props} onSubmit={this.onSubmit} />;
   }
+}
 
-  async onSubmit(metric, f) {
-    let { tableMetadata } = this.props;
-    if (metric.id != null) {
-      await this.props.updateMetric(metric);
-      this.props.clearRequestState({ statePath: ["entities", "metrics"] });
-      MetabaseAnalytics.trackEvent("Data Model", "Metric Updated");
-    } else {
-      await this.props.createMetric(metric);
-      this.props.clearRequestState({ statePath: ["entities", "metrics"] });
-      MetabaseAnalytics.trackEvent("Data Model", "Metric Created");
-    }
+@Tables.load({
+  id: (state, props) => parseInt(props.location.query.table),
+  wrapped: true,
+})
+class CreateMetricForm extends Component {
+  constructor() {
+    super();
+    this.onSubmit = this.onSubmit.bind(this);
+  }
 
+  async onSubmit(metric) {
+    const { id: tableId, db_id: databaseId } = this.props.table;
+    await this.props.createMetric({ ...metric, table_id: tableId });
+    MetabaseAnalytics.trackEvent("Data Model", "Metric Updated");
     this.props.onChangeLocation(
-      "/admin/datamodel/database/" +
-        tableMetadata.db_id +
-        "/table/" +
-        tableMetadata.id,
+      `/admin/datamodel/database/${databaseId}/table/${tableId}`,
     );
   }
 
   render() {
-    return (
-      <div>
-        <MetricForm {...this.props} onSubmit={this.onSubmit.bind(this)} />
-      </div>
+    if (this.props.table) {
+      this.props.table.fetchTableMetadata();
+    }
+    return <MetricForm {...this.props} onSubmit={this.onSubmit} />;
+  }
+}
+
+@connect(
+  mapStateToProps,
+  mapDispatchToProps,
+)
+export default class MetricApp extends Component {
+  render() {
+    return this.props.params.id ? (
+      <UpdateMetricForm {...this.props} />
+    ) : (
+      <CreateMetricForm {...this.props} />
```

Comment:
Previously, this was just one component that checked whether there was an `id` param in `componentWillMount` to know how to load the table and optionally metric. That didn't seem possible with the object loaders (let me know if that's not the case!), so I split it into two separate components that are wrapped with different object loaders.

Reply:
Yeah that's reasonable.